src/primitives/iter.rs

```diff
@@ -0,0 +1,441 @@
+            None => (min, max),
+        };
+
+        let min = bytes_len_to_fes_len(min);
+        let max = max.map(bytes_len_to_fes_len);
+
+        (min, max)
+    }
+}
+
+/// The number of fes encoded by n bytes, rounded up because we pad the fes.
+fn bytes_len_to_fes_len(bytes: usize) -> usize {
+    let bits = bytes * 8;
+    (bits + 4) / 5
+}
+
+/// Iterator adaptor that converts GF32 elements to bytes.
+///
+/// If the total number of bits is not a multiple of 8, any trailing bits are dropped.
+///
+/// Note that if there are 5 or more trailing bits, the result will be that an entire field element
+/// is dropped. If this occurs, the input was an invalid length for a bech32 string, but this
+/// iterator does not do any checks for this.
+#[derive(Clone, PartialEq, Eq)]
+pub struct Fe32ToByteIter<I: Iterator<Item = Fe32>> {
+    last_fe: Option<Fe32>,
+    bit_offset: usize,
+    iter: I,
+}
+
+impl<I> Iterator for Fe32ToByteIter<I>
+where
+    I: Iterator<Item = Fe32>,
+{
+    type Item = u8;
+
+    fn next(&mut self) -> Option<u8> {
+        let bit_offset = {
+            let ret = self.bit_offset;
+            self.bit_offset = (self.bit_offset + 8) % 5;
+            ret
+        };
+
+        if let Some(last) = self.last_fe {
+            let mut ret = last.0 << (3 + bit_offset);
+
+            self.last_fe = self.iter.next();
+            let next1 = self.last_fe?;
+            if bit_offset > 2 {
+                self.last_fe = self.iter.next();
+                let next2 = self.last_fe?;
+                ret |= next1.0 << (bit_offset - 2);
+                ret |= next2.0 >> (7 - bit_offset);
+            } else {
+                ret |= next1.0 >> (2 - bit_offset);
+                if self.bit_offset == 0 {
+                    self.last_fe = self.iter.next();
+                }
+            }
+
+            Some(ret)
+        } else {
+            None
+        }
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        // If the total number of bits is not a multiple of 8, any trailing bits are dropped.
+        let fes_len_to_bytes_len = |n| n * 5 / 8;
+
+        let (fes_min, fes_max) = self.iter.size_hint();
+        let min = fes_len_to_bytes_len(fes_min) + 1;
+        let max = fes_max.map(|max| fes_len_to_bytes_len(max) + 1);
+        (min, max)
+    }
+}
+
+/// Iterator adaptor for field-element-yielding iterator, which tacks a checksum onto the end of the
+/// yielded data.
+#[derive(Clone, PartialEq, Eq)]
+pub struct ChecksummedIter<I, Ck>
+where
+    I: Iterator<Item = Fe32>,
+    Ck: Checksum,
+{
+    iter: I,
+    checksum_remaining: usize,
+    checksum_engine: checksum::Engine<Ck>,
+}
+
+impl<I, Ck> ChecksummedIter<I, Ck>
+where
+    I: Iterator<Item = Fe32>,
+    Ck: Checksum,
+{
+    /// Creates a new checksummed iterator which adapts a data iterator of field elements by
+    /// appending the checksum.
+    pub fn new(data: I) -> ChecksummedIter<I, Ck> {
```

Comment:
In e2527dc19c3e8e9a4cd678a955fc5b161e175c87:

Proposal:
* `ChecksummedIter::new` takes a HRP and a witness version.
* We add a separate `new_without_hrp` method if you're trying to use these adaptors with a non-hrpstring.
* If it has an HRP/witness byte, it should yield all the Fe32s that would be fed into the checksum engine
* Then a new `fn hrpstring` would adapt this to instead output characters (so, HRP chars instead of fe32s, etc)

Reply:
Alright, this took a few iterations. Let me know what you think. I've kept at trying to enable various use cases while limiting the transitions from one adaptor to another to only valid transitions.

Reply:
I like it! Had some comments but I stuck them on the other thread so I'll resolve this one.